httpclient/api_client.go

```diff
@@ -0,0 +1,386 @@
+	cfg.HTTPTimeout = time.Duration(orDefault(int(cfg.HTTPTimeout), int(30*time.Second)))
+	rateLimiter := rate.NewLimiter(rate.Limit(orDefault(cfg.RateLimitPerSecond, 15)), 1)
+	if cfg.ErrorMapper == nil {
+		// default generic error mapper
+		cfg.ErrorMapper = DefaultErrorMapper
+	}
+	if cfg.ErrorRetriable == nil {
+		// by default, we just retry on HTTP 429/504
+		cfg.ErrorRetriable = DefaultErrorRetriable
+	}
+	return &ApiClient{
+		config:      cfg,
+		rateLimiter: rateLimiter,
+		httpClient: &http.Client{
+			Timeout:   cfg.HTTPTimeout,
+			Transport: cfg.httpTransport(),
+		},
+	}
+}
+
+type ApiClient struct {
+	config      ClientConfig
+	rateLimiter *rate.Limiter
+	httpClient  *http.Client
+}
+
+type DoOption struct {
+	in   RequestVisitor
+	out  func(body *responseBody) error
+	body any
+}
+
+// Do sends an HTTP request against path.
+func (c *ApiClient) Do(ctx context.Context, method, path string, opts ...DoOption) error {
+	visitors := c.config.Visitors[:]
+	for _, o := range opts {
+		if o.in == nil {
+			continue
+		}
+		// merge client-wide and request-specific visitors
+		visitors = append(visitors, o.in)
+	}
+	var requestBody any
+	for _, o := range opts {
+		if o.body == nil {
+			continue
+		}
+		requestBody = o.body
+	}
+	responseBody, err := c.perform(ctx, method, path, requestBody, visitors...)
+	if err != nil {
+		return err
+	}
+	for _, o := range opts {
+		if o.out == nil {
+			continue
+		}
+		err = o.out(responseBody)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *ApiClient) fromResponse(r *http.Response) (responseBody, error) {
+	if r == nil {
+		return responseBody{}, fmt.Errorf("nil response")
+	}
+	if r.Request == nil {
+		return responseBody{}, fmt.Errorf("nil request")
+	}
+	streamResponse := r.Request.Header.Get("Accept") != "application/json" && r.Header.Get("Content-Type") != "application/json"
+	if streamResponse {
+		return newResponseBody(r.Body, r.Header)
+	}
+	defer r.Body.Close()
+	bs, err := io.ReadAll(r.Body)
+	if err != nil {
+		return responseBody{}, fmt.Errorf("response body: %w", err)
+	}
+	return newResponseBody(bs, r.Header)
+}
+
+func (c *ApiClient) redactedDump(prefix string, body []byte) (res string) {
+	return bodyLogger{
+		debugTruncateBytes: c.config.DebugTruncateBytes,
+	}.redactedDump(prefix, body)
+}
+
+func (c *ApiClient) isRetriable(ctx context.Context, err error) bool {
+	if c.config.ErrorRetriable(ctx, err) {
+		return true
+	}
+	_, isIO := err.(*url.Error)
+	if isIO {
+		// all IO errors are retriable
+		logger.Debugf(ctx, "Attempting retry because of IO error: %s", err)
+		return true
+	}
+	message := err.Error()
+	// Handle transient errors for retries
+	for _, substring := range c.config.TransientErrors {
+		if strings.Contains(message, substring) {
+			logger.Debugf(ctx, "Attempting retry because of %#v", substring)
+			return true
+		}
+	}
+	// some API's recommend retries on HTTP 500, but we'll add that later
```

Comment:
where in this PR that maps HTTP error code to retryable/not retryable?
(although generally I guess it's accepted that 500 means no retries.. must be some important legacy API)